typescript/looker/rtl/oauthSession.spec.ts

```diff
@@ -0,0 +1,180 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+import { OAuthSession } from "./oauthSession";
+import {ApiSettings, DefaultSettings, IApiSettings} from "./apiSettings";
+import {BrowserCryptoHash, BrowserTransport} from "./browserTransport";
+import {NodeCryptoHash, NodeTransport} from "./nodeTransport";
+import {BrowserServices} from "./browserServices";
+import {MockCrypto} from "./mocks";
+import Url from "url";
+import {ITransport} from "./transport";
+
+describe('oauthSession', () => {
+  const settings = {
+    ...DefaultSettings(),
+    client_id: '123456',
+    redirect_uri: 'https://myapp.com/redirect',
+    base_url: 'https://looker.com:19999/api/3.1/',
+    looker_url: 'https://looker.com:9999',
+  } as IApiSettings
+
+  it('fails if missing settings', () => {
+    for (let key of ['client_id','redirect_uri','base_url','looker_url']) {
+      let dup: any = {...settings}
+      delete dup[key]
+      const svcs = new BrowserServices({settings: dup})
+      expect(() => new OAuthSession(svcs)).toThrow(
+      `Missing required configuration setting: '${key}'`,
+      )
+    }
+  })
+
+  it('defaults to BrowserTransport and BrowserCryptoHash', () => {
+    const services = new BrowserServices({settings: settings})
+    const session = new OAuthSession(services)
+    expect(session.transport).toBeInstanceOf(BrowserTransport)
+    expect(session.crypto).toBeInstanceOf(BrowserCryptoHash)
+  })
+
+  it('accepts transport and crypto overrides via services', () => {
+    const transport = new NodeTransport(settings)
+    const crypto = new MockCrypto()
+    const services = new BrowserServices(
+    {
+      settings: settings,
+      transport: transport,
+      crypto: crypto,
+    }
+    )
+    const session = new OAuthSession(services)
+    expect(session.crypto).toEqual(crypto)
+
+    const session2 = new OAuthSession({
+      settings: settings,
+      transport: transport,
+      crypto: crypto,
+    })
+    expect(session2.transport).toEqual(transport)
+    expect(session2.crypto).toEqual(crypto)
+  })
+
+  // No need to test for property omission via duck type hash, since the
+  // services param is not a partial, TypeScript will error for any hash
+  // that does not provide all properties of IPlatformServices
+  it('accepts transport and crypto overrides via duck type hash', () => {
+    const transport = new NodeTransport(settings)
+    const crypto = new MockCrypto()
+    const session = new OAuthSession({
+      settings: settings,
+      transport: transport,
+      crypto: crypto,
+    })
```

Comment:
Minor code tip. In Typescript, you could do:

```typescript
const session = new OAuthSession({
  settings,
  transport,
  crypto,
})
```

Reply:
👍